sources/packages/api/src/route-controllers/education-program-offering/education-program-offering.controller.service.ts

```diff
@@ -108,4 +134,152 @@ export class EducationProgramOfferingControllerService {
+   * @param programId program id.
+   * @param payload information to generate the model to perform the offering
+   * validation and persistence.
+   * @returns offering model to be validated and saved.
+   */
+  async mapOfferingAPIInDTOTOSaveOfferingModel(
+    institutionId: number,
+    locationId: number,
+    programId: number,
+    payload: EducationProgramOfferingAPIInDTO,
+  ): Promise<SaveOfferingModel> {
+    // Program information required to perform the offering validations.
+    const program = await this.programService.getEducationProgramDetails(
+      programId,
+      institutionId,
+    );
+    if (!program) {
+      throw new NotFoundException(
+        "Program to create the offering not found for the institution.",
+      );
+    }
+    return {
+      ...payload,
+      locationId,
+      studyBreaks: payload.breaksAndWeeks?.studyBreaks,
+      programContext: program,
+    };
+  }
+
+  /**
+   * Verify if all CSV validations were performed with success and throw
+   * a BadRequestException in case of some failure.
+   * @param csvValidations validations to be verified.
+   */
+  assertCSVValidationsAreValid(csvValidations: OfferingCSVValidationResult[]) {
+    const csvValidationsErrors = csvValidations.filter(
+      (csvValidation) => csvValidation.errors.length,
+    );
+    if (csvValidationsErrors.length) {
+      // At least one error was detected and the CSV must be fixed.
+      const validationResults: OfferingBulkInsertValidationResultAPIOutDTO[] =
+        csvValidationsErrors.map((validation) => ({
+          recordIndex: validation.index,
+          locationCode: validation.csvModel.institutionLocationCode,
+          sabcProgramCode: validation.csvModel.sabcProgramCode,
+          startDate: validation.csvModel.studyStartDate,
+          endDate: validation.csvModel.studyEndDate,
+          errors: validation.errors,
+          warnings: [],
+        }));
+      throw new BadRequestException(
+        new ApiProcessError(
+          "Some CSV data field received is not in correct format.",
+          OFFERING_VALIDATION_CSV_FORMAT_ERROR,
+          validationResults,
+        ),
+      );
+    }
+  }
+
+  /**
+   * Verify if all offerings validations were performed with success and throw
+   * a UnprocessableEntityException in case of some failure.
+   * @param offeringValidations validations to be verified.
+   * @param csvModels reference CSV models to provide extra context in case
+   * an error must be generated.
+   */
+  assertOfferingsValidationsAreValid(
+    offeringValidations: OfferingValidationResult[],
+    csvModels: OfferingCSVModel[],
+  ) {
+    const offeringValidationsErrors = offeringValidations.filter(
+      (offering) => offering.errors.length,
+    );
+    if (offeringValidationsErrors.length) {
+      // There is some critical validation error that will prevent some offering to be inserted.
+      const validationResults: OfferingBulkInsertValidationResultAPIOutDTO[] =
```

Comment:
Line 215-235 is almost similar to the assertCSVValidationsAreValid. Is it possible to combine them into one method?

Reply:
They do have a similar structure but there are acting on different objects. To allow them to be centralized we need to create some common interfaces between the objects `OfferingCSVValidationResult` and `OfferingValidationResult` for instance. I am not sure if the effort is valid for this case. We can talk in the meeting about it.